### PR TITLE
Adding SectionName to PolicyTargetReference

### PR DESCRIPTION
**Pull request by zhaohuabing**
This allows attaching policies to a specific section within the targeted object.

Related issue: https://github.com/kubernetes-sigs/gateway-api/issues/2147

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

What this PR does:

This PR allows attaching policies to a specific section within the targeted object.

Why we need it:

Lacking SectionName in PolicyTargetReference is blocking my current work. I'm implementing a direct policy attachment and need to target a specific listener inside a gateway. 

```release-note
NONE
```

### CODE REVIEW

**Comment by robscott** [comment]
Hey @zhaohuabing, thanks for working on this! If I'm understanding correctly, it looks like you're trying to take on a couple of separate issues:

1. Adding SectionName to PolicyTargetReference (https://github.com/kubernetes-sigs/gateway-api/issues/2147)
2. Adding Name to Route Rules (https://github.com/kubernetes-sigs/gateway-api/issues/995)

I think we mostly have consensus on 1, but we're waiting for #2128 to merge before moving forward. On 2, I don't think we quite have enough consensus to move forward yet. Maybe once #2128 merges you can check in with @arkodg to see if he needs any help with the implementation of #2147.

It may also be helpful to take a look at our [documentation for the GEP process](https://gateway-api.sigs.k8s.io/geps/overview/). All API changes have to go through that process so we can't start directly with a PR to change the API itself without first having an approved GEP in an "implementable" state.

**Comment by zhaohuabing** [comment]
@robscott What I really need is to have a `SectionName` field in the `PolicyTargetReference API`. Right now, its absence is causing a bit of a roadblock for me as I'm trying to set up a direct policy attachment and want to target a specific listener within a gateway.  

Since the community have already reached agreement on 1, could we consider adding this field as a first step? This matter is quite time-sensitive for me, and I'm a bit worried that it will take a while to get https://github.com/kubernetes-sigs/gateway-api/pull/2128 merged. Moreover, PR 2128 is mainly about the discoverability and doesn't related to the `SectionName` field at all, so it shouldn't block adding this field.

**Comment by arkodg** [comment]
hey @zhaohuabing can you also edit https://github.com/kubernetes-sigs/gateway-api/blob/4651e916c370af99c87312de647adb00fe176b2a/geps/gep-713.md?plain=1#L685 to now account for `SectionName` . Imo, since this PR is in draft mode, it shouldn't distract others from reviewing the Policy Discoverability PR first

**Comment by zhaohuabing** [comment]
Change to "ready to review". Since https://github.com/kubernetes-sigs/gateway-api/pull/2128 has already been merged, we can move on with this.

File: apis/v1alpha2/policy_types.go
Comment (arkodg) [inline apis/v1alpha2/policy_types.go:46]: Route Rule Name doesnt exist in the spec yet, so should be added only after it has been formally added into the spec

File: apis/v1alpha2/policy_types.go
Comment (youngnick) [inline apis/v1alpha2/policy_types.go:43]:
```suggestion
	// unspecified, this targetRef targets the entire resource. In the following
```
What "this" refers to is a little unclear, better to be specific.

File: geps/gep-713.md
Comment (youngnick) [inline geps/gep-713.md:1269]:
```suggestion
The `sectionName` field can also be used to target a specific section of other resources, for example:
```
"Could" is a holdover from the previous language and is not as definite as "can".

File: geps/gep-713.md
Comment (youngnick) [inline geps/gep-713.md:1271]: I'd move this one to below `Service.Ports`, since Service is already defined.

**Review by youngnick — COMMENTED** [review]
This is looking great, thanks @zhaohuabing. Just a few small things to change.

File: geps/gep-713.md
Comment (zhaohuabing) [reply geps/gep-713.md:1269]: Thanks! As a non-native speaker, I'm still trying to improve my English writing :-).

File: apis/v1alpha2/policy_types.go
Comment (youngnick) [inline apis/v1alpha2/policy_types.go:47]: On further review, I think we need to be specific here about what happens if this is not found.

Would you be able to add some wording like:
```go
	// 
	// If a SectionName is specified, but does not exist on the targeted object,
	// the Policy must fail to attach, and the policy implementation should record
	// a `ResolvedRefs` or similar Condition in the Policy's status.
```

(We should specify something similar at the top level as well, but that's another issue).

File: geps/gep-713.md
Comment (youngnick) [inline geps/gep-713.md:1275]: If you could add a section summarizing what happens if the SectionName doesn't exist here as well please, that would be great.

**Comment by youngnick** [comment]
Yep, this LGTM now. I'll approve and hold for other review though.

/approve
/hold

**Review by shaneutt — APPROVED** [review]
I think this looks helpful, but I think given the importance of policy attachment stuff with GA looming Rob should take a final look here as well.

/cc @robscott

**Comment by robscott** [comment]
Thanks for working on this @zhaohuabing! Our policy attachment model is already quite complex so I want to ensure we've really thought out any potential addition before moving forward. Some questions that I'd like to see answered in the GEP:

1. How does this interact with inherited policy attachment? [Our chart here is already pretty complicated](https://gateway-api.sigs.k8s.io/geps/gep-713/#hierarchy), and as far as I can tell, this would nearly double the size of it by adding new "section" levels for nearly every level.
2. What are the specific use cases we're adding this for? I know Gateway -> Backend TLS may benefit from this, is there anything else?
3. In the case of Gateway -> Backend TLS it's possible that `port` would be a more natural extension than `sectionName`, adding both could result in an absolute explosion of complexity for policy attachment, especially "inherited" policy attachment.
4. The current PR adds `sectionName` to the existing PolicyTargetRef type, should we instead add a separate type so some implementations can choose to create Policy resources that intentionally do not support SectionName?
5. If this feature ends up getting paired with a name on Route rules, when would an implementation use a custom filter (ExtensionRef) instead of a Policy? (I'm assuming this is one of the long term goals here, but let me know if I'm wrong).
6. Is there anything we can do to reduce the added complexity here? We've relied pretty heavily on the idea that tooling and systems could generically visualize policy attachment (x-ref https://github.com/michaelvl/gateway-api-lens and https://github.com/gauravkghildiyal/gwctl), how would those integrate with the idea of SectionName? Is there a way for implementations to document the kinds of TargetRefs they support for a resource, and more specifically, where they support SectionName and what it means?

If any of these questions are too complicated, we can split up the work to answer them and merge some updates to the GEP with outstanding questions marked as unresolved, I just want to be very careful when adding complexity to this model.

**Comment by arkodg** [comment]
hey @robscott, taking a stab at these questions

> 1.How does this interact with inherited policy attachment? [Our chart here is already pretty complicated](https://gateway-api.sigs.k8s.io/geps/gep-713/#hierarchy), and as far as I can tell, this would nearly double the size of it by adding new "section" levels for nearly every level.

* There is already language in the GEP around this
```
For Policies that affect an object's existing fields, multiple instances of the same Policy Kind affecting an object's fields will be evaluated as though only a single Policy "wins" the right to affect each field. This operation is performed on a per-distinct-field basis.
```

* Even without this change, the GEP could be expanded with examples on how multiple Inherited PolicyAttachment resources (lets say A, B, C ) affect outcome when applied to the same target resource (lets say Gateway G)

* For now, we can also add language in this PR, that says - `current recommendation is for Direct Policy Attachment types only` to limit scope and enhance in a future PR

> 2. What are the specific use cases we're adding this for? I know Gateway -> Backend TLS may benefit from this, is there anything else?

* Here's some of the many features that can be applied per Gateway Listener tied mostly to Downstream clients
* Authentication (JWT etc) ## used in the GEP as an example
* RateLimiting
* CORS
* Proxy Protocol

There are so many more that are relevant per Route Rule (tied to upstream) like Load Balancing, Retries, Timeouts, OutlierDetection, CircuitBreaker, TLS that can only be fine tuned if we add a `Name` to a route rule which doesnt exist today so the policy applied here will be coarse grained at the Route level.

> 3. In the case of Gateway -> Backend TLS it's possible that port would be a more natural extension than sectionName, adding both could result in an absolute explosion of complexity for policy attachment, especially "inherited" policy attachment.

* This PR follows the same design pattern as the [SectionName](https://github.com/kubernetes-sigs/gateway-api/blob/c3ceacfe8c302682b1f5b13a73c084cefb24926f/apis/v1beta1/shared_types.go#L117) in xRoute. This could be changed by maybe using a `Type` field making it look like
```
section:
  type: "Port"
  port: 8080
```
or
```
section:
  type: "Listener"
  listener: foo
```
but imo then similar treatment should also be done at the xRoute level 

> 4. The current PR adds sectionName to the existing PolicyTargetRef type, should we instead add a separate type so some implementations can choose to create Policy resources that intentionally do not support SectionName?

* sure this works, but for the case where an implementation doesnt use it, it could either ignore the field or use the `Status` to highlight the fact that it has ignored the field

> 5. If this feature ends up getting paired with a name on Route rules, when would an implementation use a custom filter (ExtensionRef) instead of a Policy? (I'm assuming this is one of the long term goals here, but let me know if I'm wrong).

* Here's how we think of policy vs ext filter in Envoy Gateway, and these opinions may vary across implementations
  * if the same resource can be applied to Gateway and xRoute, Policy makes sense here
  * if the resource only applies to xRoute and is authored by app team, ext filter makes sense here
  * if the resource needs to be overlayed/enforced on top of a xRoute, by another author / team, Policy makes sense here (transparent to the app team)

> 6. Is there anything we can do to reduce the added complexity here? We've relied pretty heavily on the idea that tooling and systems could generically visualize policy attachment (x-ref https://github.com/michaelvl/gateway-api-lens and https://github.com/gauravkghildiyal/gwctl), how would those integrate with the idea of SectionName? Is there a way for implementations to document the kinds of TargetRefs they support for a resource, and more specifically, where they support SectionName and what it means?

both these tools will be useful in helping a user visualize, which policy affects which resource/section/field.
In Envoy Gateway, we rely on the Policy Status to also provide more info to the user

please let us know which questions you'd like to see explicit verbiage, in the GEP, thanks

**Comment by youngnick** [comment]
The more I think about @robscott's feedback here, the more I like it. I propose we do the following changes:
* We mandate that section names can only be used for Direct Policy, not Inherited. (The interactions are just too complicated if you allow sectionName + inheritance).
* We create a `DirectTargetRef` struct, that inlines the existing TargetRef struct, and adds `SectionName`.
* We update GEP-713 to with details about SectionName, and how it only works with Direct Attached Policy.

Note that the nice part here is that if we find we need it later, we can walk this back. But this allows us to have sub-object policy targeting for now, while keeping the scope and implications manageable.

I think that adding a `type` field is too far - I think that every field where we might want to do subsection binding already has a name field (Service ports have `name` already, for example).

**Comment by youngnick** [comment]
Back to @robscott for another review, I think.

File: geps/gep-713.md
Comment (sanjaypujare) [inline geps/gep-713.md:1265]: Just curious why the example was changed from an HTTPRoute to Gateway (and from RetryPolicy to AuthenticationPolicy). Whatever is the reason, could we have both examples? Specially because you also have an example of adding a name to RouteRule (on line 1277)

Comment (zhaohuabing) [reply geps/gep-713.md:1265]: @sanjaypujare The first example is changed to Gateway because the current use case I want to address is the Gateway Listener, and we already has the name field in the Listener object.  The second example is used to demonstrate what we need to change(adding the name field) to apply policy to RouteRule as well.

Comment (sanjaypujare) [reply geps/gep-713.md:1265]: Thanks. With this change you included the RetryPolicy but not the HTTPRoute in the original doc which has the named section. From https://gateway-api.sigs.k8s.io/geps/gep-713/#apply-policies-to-sections-of-a-resource-future-extension I have the snippet as follows:

```
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: http-app-1
  labels:
    app: foo
spec:
  hostnames:
  - "foo.com"
  rules:
  - name: bar
    matches:
    - path:
        type: Prefix
        value: /bar
    forwardTo:
    - serviceName: my-service1
      port: 8080
---
```

This is just above the RetryPolicy so it will be good to include (or rather restore) the above snippet before line 1277.

**Review by sanjaypujare — APPROVED** [review]
LGTM

**Review by sanjaypujare — APPROVED** [review]

File: apis/v1alpha2/policy_types.go
Comment (youngnick) [inline apis/v1alpha2/policy_types.go:45]: Small nit, since it's not only Gateway API resources you can target.
```suggestion
// should be used as part of Policy resources that can target single
// resources. For more information on how this policy attachment model works,
```

File: geps/gep-713.md
Comment (youngnick) [inline geps/gep-713.md:1342]: This is great, but we will probably also need to update the parts about Direct Attachment further up in this document with the details that you should use the DirectTargetRef instead of the more generic one.

(Non-blocking, just a note for myself for the next time I touch this GEP).

**Review by youngnick — COMMENTED** [review]
This LGTM, I have no more objections.

**Comment by arkodg** [comment]
a non specific vs specific `targetRef` applied to the same resource for a Direct Policy Attachment, does introduce the notion of overrides e.g. if `PolicyA` is applied to 
```
 targetRef:
     name: foo-gateway
     group: gateway.networking.k8s.io
     kind: Gateway
```

and if `PolicyB` is applied to 

```
 targetRef:
     name: foo-gateway
     group: gateway.networking.k8s.io
     kind: Gateway
     name: some-listener
```

what is the final setting for `foo-gateway.some-listener` ?
* implementation specific. # safest answer to add in this GEP
* most specific target wins

**Comment by sanjaypujare** [comment]
>...
>
> * all settings from policy with most specific target wins

I vote for that.

**Comment by zhaohuabing** [comment]
> * all settings from policy with most specific target wins

I prefer this one because it's simple and clear.

> * some settings from policy with most specific target are used as overrides and least specific policy is used as default for the rest of the unset settings

This one seems right but it would be very hard for users to figure out which settings are from which policy.

**Comment by youngnick** [comment]
Yes, we use "most specific wins" in other places, so I am very much in favor.

@zhaohuabing, sorry, but you'll need to include some information in the SectionName part of GEP-713 about this behavior. Something like:

When multiple Policies of the same type target the same object, one with a SectionName specified, and one without, the one with a SectionName is more specific, and so will have all its settings apply. The less-specifc Policy will not attach to the target.

I'll also take a note to add or extend conflict-resolution rules for direct policy attachment further up, with the "more specific wins" rule. (I have more updates to do to GEP-713 already, this will add to the pile).

**Comment by youngnick** [comment]
Thanks for that last change @zhaohuabing, this LGTM.

/lgtm

**Comment by youngnick** [comment]
Ping @robscott for one more check, then we can remove the hold.

**Comment by youngnick** [comment]
@zhaohuabing, you'll need to rerun `make generate` and commit the result for CI to pass.

**Review by robscott — COMMENTED** [review]
Thanks for all the work on this @zhaohuabing! As much as I'm hesitant to add additional complexity here, this does feel inevitable. Have a few nits, but this looks good overall. In general just trying to at least make sure we have some guidance in here that ensures this is only used when it's actually necessary.

Would also appreciate some feedback from @kflynn on this because they've raised some important concerns about the complexity of our existing policy attachment model, so want to be careful when adding more.

File: apis/v1alpha2/policy_types.go
Comment (zhaohuabing) [reply apis/v1alpha2/policy_types.go:51]: I added comment below and hope it clarify the use of `DirectPolicyTargetReference`:

```
// Note: This should only be used for direct policy attachment when references
// to SectionName are actually needed. In all other cases, PolicyTargetReference
// should be used.
```

Comment (youngnick) [reply apis/v1alpha2/policy_types.go:51]: I love the comment, but I agree that the naming doesn't help with possible confusion here. I think we should go with clarity over brevity here and call the thing `PolicyTargetReferenceWithSectionName`. Pretty long, but does exactly what it says.

**Comment by robscott** [comment]
Thanks @zhaohuabing!

/lgtm
/hold cancel

**Pull request merged**
